apache/superset
# feat(explore): Show confirmation modal if user exits Explore without saving changes

### SUMMARY
This PR adds a browser confirmation modal (similar to the one that we have in dashboard edit mode) when user makes a change in control panel and then tries to close Explore (either by closing the tab or moving to another page) without clicking Save button.
The modal that displays when user navigates out of SPA context (some external page or a Superset page that hasn't been migrated to SPA yet) or closes the tab, is not customizable.
In the case of navigating within SPA context, we can easily display a custom message in the modal thanks to react-router's `Prompt` component. However, displaying a custom modal instead of the default one rendered by `<Prompt />` requires more effort and will be handled in a separate PR.

CC @kasiazjc 

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF


https://user-images.githubusercontent.com/15073128/188621562-23b4e1cf-32e7-440a-9b03-ae90cdf398d4.mov


### TESTING INSTRUCTIONS
1. Open a chart and make some changes in control panel
2. Try to go to another SPA page (like charts list)
3. A modal should appear. Clicking Leave should discard your changes and go to next page. Clicking Cancel should close the modal.
4. Make some changes in control panel and try to close browser tab or go to some external page.
5. A default browser confirmation modal should appear (it's not customizable)
6. Save your changes
7. Now leaving the page should not trigger the confirmation modal

KNOWN LIMITATION: if the previous page is a SPA route (for example you opened Explore from charts list), clicking browser back button will not trigger the confirmation modal. It's very bugged in react-router and I decided to disable it for now. We'll try to find a workaround in the future

### ADDITIONAL INFORMATION
